artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java

```diff
@@ -559,6 +587,10 @@ public void preAcknowledge(final Transaction tx, final MessageReference ref, fin
          return;
       }
 
+      if (isBlockedByNoForward()) {
+         return;
+      }
+
```

Comment:
I suspect a different check will be needed here, inspecting the specific message; this is ultimately called by the queue, which I expect usually wont have a mirror target controller set. It will also want to accommodate not sending acks for no-forward messages [that it didnt send to begin with] even across broker restarts, due to expiry, etc etc.

I expect the test-peer based testing I know you are working on will demonstrate this.

Reply:
I've implemented some tests and they seem to work despite the assumption we had that it would not. Maybe something is worth a double check?

Reply:
I think your changes last week have actually already stopped it sending some of the stuff I previously thought it would and checked that it actually did.

There are other scenarios that are less clear though. For example, if you consume the message from the middle broker instead of the first, its going to mirror the ack onward, even though it didnt mirror the message originally, which is marked as no-forward.

Reply:
I believe this is addressed now, feel free to reopen if you catch a mistake.